client/updates/updates.go

```
@@ -0,0 +1,110 @@
+package updates
+
+import (
+	"sync"
+	"time"
+
+	chshare "github.com/cloudradar-monitoring/rport/share"
+)
+
+type PackageManager interface {
+	Detect() bool
+	GetUpdatesStatus() *Status
+}
+
+type Updates struct {
+	status *Status
+	mtx    sync.Mutex
+
+	interval time.Duration
+	stop     chan struct{}
+
+	pkgMgr PackageManager
+	logger *chshare.Logger
+}
+
+type Status struct {
+	Refreshed                time.Time
+	UpdatesAvailable         int
+	SecurityUpdatesAvailable int
+	UpdateSummaries          []Summary
+	RebootPending            bool
+	Error                    string
+	Hint                     string
+}
+
+type Summary struct {
+	Title            string
+	Description      string
+	RebootRequired   bool
+	IsSecurityUpdate bool
+}
+
+func New(logger *chshare.Logger, interval time.Duration) *Updates {
+	u := &Updates{
+		interval: interval,
+		stop:     make(chan struct{}),
+
+		logger: logger,
+	}
+
+	packageManagers := []PackageManager{
+		NewAptPackageManager(),
+	}
+	for _, pm := range packageManagers {
+		if pm.Detect() {
+			u.pkgMgr = pm
```

what if I have an ubuntu without a package manager? I think this is an error case but this code silently ignores such cases. As we have knowledge about which os our clients have we can try to use a  package manager specific to the client's os rather than try to execute all possible package managers against all possible OS versions

---

Running ubuntu without a package manager should be a valid option, in that case the detection will fail and we will report that there is no supported package manager.
I think it's simpler to just detect if each package manager is available rather than construct a complex mapping from all possible OS versions to their package manager (a mapping we would have to maintain and update in the future). There are probably over 10 distributions that use apt and new ones could be developed at any time.